### PR TITLE
LibCodeComprehension: Re-organize code comprehension related code

### PR DESCRIPTION
**Pull request by itamar8910**
This moves all code comprehension-related code our of HackStudio and into a new library, LibCodeComprehension.

This change will enable us to make use of code comprehension in other parts of the project, for example for code generation.

(Sorry for the single large commit, it's mostly moving things around & adjusting namespaces).

### CODE REVIEW

File: Userland/Libraries/LibCodeComprehension/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+set(SOURCES
+        CodeComprehensionEngine.cpp
+        FileDB.cpp
+        )
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCodeComprehension/CMakeLists.txt:4]: that should be unindented

File: Userland/Libraries/LibCodeComprehension/CodeComprehensionEngine.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+class CodeComprehensionEngine {
+    AK_MAKE_NONCOPYABLE(CodeComprehensionEngine);
+    AK_MAKE_NONMOVABLE(CodeComprehensionEngine);
+
+public:
+    CodeComprehensionEngine(FileDB const& filedb, bool store_all_declarations = false);
+    virtual ~CodeComprehensionEngine() = default;
+
+    virtual Vector<AutocompleteResultEntry> get_suggestions(String const& file, const GUI::TextPosition& autocomplete_position) = 0;
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCodeComprehension/CodeComprehensionEngine.h:27]: While you're moving things around, how about east const?

File: Userland/Libraries/LibCodeComprehension/Cpp/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
+set(SOURCES
+        CppComprehensionEngine.cpp
+        )
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCodeComprehension/Cpp/CMakeLists.txt:3]: Same here

File: Userland/Libraries/LibCodeComprehension/Cpp/Tests.cpp [hunk 4]
```diff
@@ -9,9 +9,7 @@
 #include "CppComprehensionEngine.h"
 #include <AK/LexicalPath.h>
 #include <LibCore/File.h>
-
-using namespace LanguageServers;
-using namespace LanguageServers::Cpp;
+#include <Libraries/LibMain/Main.h>
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCodeComprehension/Cpp/Tests.cpp:12]: Why is there a `Libraries`?

File: Userland/Libraries/LibCodeComprehension/Shell/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+set(SOURCES
+        ShellComprehensionEngine.cpp
+        )
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCodeComprehension/Shell/CMakeLists.txt:3]: same here

File: Userland/Libraries/LibCodeComprehension/Types.h [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+#include <AK/String.h>
+
+#pragma once
+
+namespace CodeComprehension {
+
+enum class Language {
+    Unspecified,
+    Cpp,
+};
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCodeComprehension/Types.h:16]: Why is there no Shell here?

File: Userland/Libraries/LibCodeComprehension/Types.h [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    DeclarationType type;
+    String scope;
+
+    bool operator==(Declaration const& other) const
+    {
+        return name == other.name && position == other.position && type == other.type && scope == other.scope;
+    }
+};
+
+#define FOR_EACH_SEMANTIC_TYPE        \
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCodeComprehension/Types.h:64]: Let's have the other macro as an argument to this one; I think that pattern is more common.

**Review by kleinesfilmroellchen — COMMENTED** [review]
As a fellow connoisseur of refactoring, I very much approve.

File: Userland/Libraries/LibCodeComprehension/Types.h (hunk 6 above)
Comment (itamar8910) [reply Userland/Libraries/LibCodeComprehension/Types.h:16]: Turns out we leave `AutocompleteResultEntry::language` as `Unspecified` in many places, and yeah we don't use it at all for Shell stuff.

It doesn't look like something that should be fixed as part of this PR though, so I added a TODO for that :)

File: Userland/Libraries/LibCodeComprehension/Types.h (hunk 7 above)
Comment (itamar8910) [reply Userland/Libraries/LibCodeComprehension/Types.h:65]: Not sure I understand what you meant here. Can you point at an example in the repo for doing it the other way around?

Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibCodeComprehension/Types.h:65]: okay nvm

**Pull request merged**